lib/mutgen.c

```diff
@@ -183,69 +198,94 @@ mutgen_init_allocator(mutgen_t *self, tsk_table_collection_t *tables)
 
 static int MSP_WARN_UNUSED
 mutgen_add_mutation(mutgen_t *self, node_id_t node, double position,
-        const char *ancestral_state, const char *derived_state)
+        const char *ancestral_state, const char *derived_state,
+        avl_node_t* avl_node)
 {
     int ret = 0;
-    tsk_site_t *site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
+    bool new_site = (avl_node == NULL);
+    site_node_t *site_node;
+    tsk_site_t *site;
+    mutation_node_t *mutation_node = tsk_blkalloc_get(&self->allocator, sizeof(*mutation_node));
     tsk_mutation_t *mutation = tsk_blkalloc_get(&self->allocator, sizeof(*mutation));
-    avl_node_t* avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
-
-    if (site == NULL || mutation == NULL || avl_node == NULL) {
+    if (mutation == NULL || mutation_node == NULL) {
         ret = MSP_ERR_NO_MEMORY;
         goto out;
     }
-    memset(site, 0, sizeof(*site));
+    if (new_site) {
+        avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
+        site_node = tsk_blkalloc_get(&self->allocator, sizeof(*site_node));
+        site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
+        if (site == NULL || site_node == NULL || avl_node == NULL) {
+            ret = MSP_ERR_NO_MEMORY;
+            goto out;
+        }
+        memset(site_node, 0, sizeof(*site_node));
```

Comment:
Probably don't need the memset here for the list node, as we're setting all the members anyway. It doesn't do any harm though.